glusterfsd/src/glusterfsd.c

```diff
@@ -1482,6 +1482,142 @@ cleanup_and_exit(int signum)
     }
 }
 
+static FILE *
+get_volfp(glusterfs_ctx_t *ctx)
+{
+    cmd_args_t *cmd_args = NULL;
+    FILE *specfp = NULL;
+
+    cmd_args = &ctx->cmd_args;
+
+    if ((specfp = fopen(cmd_args->volfile, "r")) == NULL) {
+        gf_smsg("glusterfsd", GF_LOG_ERROR, errno, glusterfsd_msg_9,
+                "volume_file=%s", cmd_args->volfile, NULL);
+        return NULL;
+    }
+
+    gf_msg_debug("glusterfsd", 0, "loading volume file %s", cmd_args->volfile);
+
+    return specfp;
+}
+
+static int
+volfile_init(glusterfs_ctx_t *ctx)
+{
+    int ret = -1;
+    char *volfile = NULL;
+    gf_volfile_t *volfile_obj = NULL;
+    gf_volfile_t *volfile_tmp = NULL;
+    struct stat stbuf = {
+        0,
+    };
+    char sha256_hash[SHA256_DIGEST_LENGTH] = {
+        0,
+    };
+
+    cmd_args_t *cmd_args = &ctx->cmd_args;
+    FILE *fp = get_volfp(ctx);
+    if (!fp) {
+        gf_smsg("glusterfsd", GF_LOG_ERROR, 0, glusterfsd_msg_28, NULL);
+        goto out;
+    }
+    ret = sys_stat(cmd_args->volfile, &stbuf);
+    if (IS_ERROR(ret)) {
+        gf_smsg("glusterfsd", GF_LOG_ERROR, errno, glusterfsd_msg_9,
+                "volume_file=%s", cmd_args->volfile, NULL);
+        goto out;
+    }
+
+    volfile = GF_MALLOC(stbuf.st_size, gf_common_mt_char);
+    if (!volfile) {
+        gf_smsg("glusterfsd", GF_LOG_ERROR, ENOMEM, glusterfsd_msg_9,
+                "volume_file=%s", cmd_args->volfile, NULL);
+        ret = -1;
+        goto out;
+    }
+    ret = fread(volfile, stbuf.st_size, 1, fp);
+    if (IS_ERROR(ret)) {
+        gf_smsg("glusterfsd", GF_LOG_ERROR, errno, glusterfsd_msg_9,
+                "volume_file=%s", cmd_args->volfile, NULL);
+        goto out;
+    }
+    glusterfs_compute_sha256((const unsigned char *)volfile, stbuf.st_size,
+                             sha256_hash);
+    LOCK(&ctx->volfile_lock);
+    {
+        list_for_each_entry(volfile_obj, &ctx->volfile_list, volfile_list)
+        {
+            if (!memcmp(sha256_hash, volfile_obj->volfile_checksum,
+                        sizeof(volfile_obj->volfile_checksum))) {
+                UNLOCK(&ctx->volfile_lock);
+                ret = 0;
+                gf_smsg(THIS->name, GF_LOG_INFO, 0, glusterfsd_msg_40, NULL);
+                goto out;
+            }
+            volfile_tmp = volfile_obj;
+            break;
```

Comment:
We only check the hash against the first item of the list. Is that on purpose ? what if the hash matches another item in the list ?

Reply:
The list will be more than 1 object only in case of brick-mux, which is why, when we start the brick process with volfile-path this is going to work without issues. Will add a comment about it here, for future reference.

Reply:
Understood.